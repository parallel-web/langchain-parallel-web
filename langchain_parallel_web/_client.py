"""Client utilities for Parallel integration."""

from __future__ import annotations

import os
from typing import Any, Optional, Union

import openai
from parallel import AsyncParallel, Parallel


def get_api_key(api_key: Optional[str] = None) -> str:
    """Retrieve the Parallel API key from argument or environment variables.

    Args:
        api_key: Optional API key string.

    Returns:
        API key string.

    Raises:
        ValueError: If API key is not found.
    """
    if api_key:
        return api_key

    env_key = os.environ.get("PARALLEL_API_KEY")
    if env_key:
        return env_key

    msg = (
        "Parallel API key not found. Please pass it as an argument or set the "
        "PARALLEL_API_KEY environment variable."
    )
    raise ValueError(msg)


def get_openai_client(api_key: str, base_url: str) -> openai.OpenAI:
    """Returns a configured sync OpenAI client for the Chat API."""
    return openai.OpenAI(api_key=api_key, base_url=base_url)


def get_async_openai_client(api_key: str, base_url: str) -> openai.AsyncOpenAI:
    """Returns a configured async OpenAI client for the Chat API."""
    return openai.AsyncOpenAI(api_key=api_key, base_url=base_url)

def _get_search_timeout(processor: str) -> float:
    """Get the timeout for the search API based on the processor type."""
    # Set timeout based on processor type:
    # - base processor: 10 seconds (typical 4-5s response time)
    # - pro processor: 90 seconds (typical 45-70s response time)
    return 90.0 if processor == "pro" else 10.0

class ParallelSearchClient:
    """Synchronous client for Parallel Search API using the Parallel SDK."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.parallel.ai",
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        # Initialize the Parallel SDK client
        self.client = Parallel(api_key=api_key, base_url=base_url)

    def search(
        self,
        objective: Optional[str] = None,
        search_queries: Optional[list[str]] = None,
        processor: str = "base",
        max_results: int = 10,
        max_chars_per_result: int = 1500,
        source_policy: Optional[dict[str, Union[str, list[str]]]] = None,
    ) -> dict[str, Any]:
        """Perform a synchronous search using the Parallel Search API via SDK."""
        if not objective and not search_queries:
            msg = "Either 'objective' or 'search_queries' must be provided"
            raise ValueError(msg)


        # Use the Parallel SDK's beta.search method
        search_response = self.client.beta.search(
            objective=objective,
            search_queries=search_queries,
            processor=processor,
            max_results=max_results,
            max_chars_per_result=max_chars_per_result,
            source_policy=source_policy,
            timeout=_get_search_timeout(processor),
        )

        # Convert the SDK response to a dictionary
        return search_response.model_dump()


class AsyncParallelSearchClient:
    """Asynchronous client for Parallel Search API using the Parallel SDK."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.parallel.ai",
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        # Initialize the Parallel SDK async client
        self.client = AsyncParallel(api_key=api_key, base_url=base_url)

    async def search(
        self,
        objective: Optional[str] = None,
        search_queries: Optional[list[str]] = None,
        processor: str = "base",
        max_results: int = 10,
        max_chars_per_result: int = 1500,
        source_policy: Optional[dict[str, Union[str, list[str]]]] = None,
    ) -> dict[str, Any]:
        """Perform an async search using the Parallel Search API via SDK."""
        if not objective and not search_queries:
            msg = "Either 'objective' or 'search_queries' must be provided"
            raise ValueError(msg)

        # Use the Parallel SDK's beta.search method
        search_response = await self.client.beta.search(
            objective=objective,
            search_queries=search_queries,
            processor=processor,
            max_results=max_results,
            max_chars_per_result=max_chars_per_result,
            source_policy=source_policy,
            timeout=_get_search_timeout(processor),
        )

        # Convert the SDK response to a dictionary
        return search_response.model_dump()


def get_search_client(
    api_key: str, base_url: str = "https://api.parallel.ai"
) -> ParallelSearchClient:
    """Returns a configured sync Parallel Search client."""
    return ParallelSearchClient(api_key, base_url)


def get_async_search_client(
    api_key: str, base_url: str = "https://api.parallel.ai"
) -> AsyncParallelSearchClient:
    """Returns a configured async Parallel Search client."""
    return AsyncParallelSearchClient(api_key, base_url)


class ParallelExtractClient:
    """Synchronous client for Parallel Extract API using the Parallel SDK."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.parallel.ai",
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        # Initialize the Parallel SDK client
        self.client = Parallel(api_key=api_key, base_url=base_url)

    def extract(
        self,
        urls: list[str],
        objective: Optional[str] = None,
        search_queries: Optional[list[str]] = None,
        excerpts: Optional[Union[bool, dict[str, Any]]] = None,
        full_content: Optional[Union[bool, dict[str, Any]]] = None,
        fetch_policy: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Perform a synchronous extract using the Parallel Extract API via SDK."""
        if not urls:
            msg = "At least one URL must be provided"
            raise ValueError(msg)

        # Set timeout based on number of URLs (5 seconds per URL)
        timeout = 5.0 * len(urls)

        # Use the Parallel SDK's beta.extract method
        extract_response = self.client.beta.extract(
            urls=urls,
            objective=objective,
            search_queries=search_queries,
            excerpts=excerpts,
            full_content=full_content,
            fetch_policy=fetch_policy,
            betas=["search-extract-2025-10-10"],
            timeout=timeout,
        )

        # Convert the SDK response to a dictionary
        return extract_response.model_dump()


class AsyncParallelExtractClient:
    """Asynchronous client for Parallel Extract API using the Parallel SDK."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.parallel.ai",
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        # Initialize the Parallel SDK async client
        self.client = AsyncParallel(api_key=api_key, base_url=base_url)

    async def extract(
        self,
        urls: list[str],
        objective: Optional[str] = None,
        search_queries: Optional[list[str]] = None,
        excerpts: Optional[Union[bool, dict[str, Any]]] = None,
        full_content: Optional[Union[bool, dict[str, Any]]] = None,
        fetch_policy: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Perform an async extract using the Parallel Extract API via SDK."""
        if not urls:
            msg = "At least one URL must be provided"
            raise ValueError(msg)

        # Set timeout based on number of URLs (5 seconds per URL)
        timeout = 5.0 * len(urls)

        # Use the Parallel SDK's beta.extract method
        extract_response = await self.client.beta.extract(
            urls=urls,
            objective=objective,
            search_queries=search_queries,
            excerpts=excerpts,
            full_content=full_content,
            fetch_policy=fetch_policy,
            betas=["search-extract-2025-10-10"],
            timeout=timeout,
        )

        # Convert the SDK response to a dictionary
        return extract_response.model_dump()


def get_extract_client(
    api_key: str, base_url: str = "https://api.parallel.ai"
) -> ParallelExtractClient:
    """Returns a configured sync Parallel Extract client."""
    return ParallelExtractClient(api_key, base_url)


def get_async_extract_client(
    api_key: str, base_url: str = "https://api.parallel.ai"
) -> AsyncParallelExtractClient:
    """Returns a configured async Parallel Extract client."""
    return AsyncParallelExtractClient(api_key, base_url)
