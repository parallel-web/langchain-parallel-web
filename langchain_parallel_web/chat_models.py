"""Parallel Web chat model integration.

This module provides the ChatParallelWeb class for interacting with Parallel's
Chat API through an OpenAI-compatible interface.
"""

from __future__ import annotations

import contextlib
from collections.abc import AsyncIterator, Iterator
from typing import Any, Optional, cast

import openai
from langchain_core.callbacks import (
    AsyncCallbackManagerForLLMRun,
    CallbackManagerForLLMRun,
)
from langchain_core.language_models import BaseChatModel
from langchain_core.messages import (
    AIMessage,
    AIMessageChunk,
    BaseMessage,
    HumanMessage,
    SystemMessage,
)
from langchain_core.outputs import ChatGeneration, ChatGenerationChunk, ChatResult
from openai import AuthenticationError, RateLimitError
from pydantic import Field, SecretStr, model_validator
from typing_extensions import Self

from ._client import get_api_key, get_async_openai_client, get_openai_client


def _convert_message_to_dict(message: BaseMessage) -> dict[str, Any]:
    """Convert a LangChain message to OpenAI message format."""
    if isinstance(message, SystemMessage):
        return {"role": "system", "content": message.content}
    if isinstance(message, HumanMessage):
        return {"role": "user", "content": message.content}
    if isinstance(message, AIMessage):
        return {"role": "assistant", "content": message.content}
    msg = f"Unsupported message type: {type(message)}"
    raise ValueError(msg)


def _prepare_messages(messages: list[BaseMessage]) -> list[dict[str, Any]]:
    """Prepare messages for API call by merging consecutive messages and converting to dict format."""  # noqa: E501
    merged_messages = _merge_consecutive_messages(messages)
    return [_convert_message_to_dict(msg) for msg in merged_messages]


def _create_response_metadata(response: Any, choice: Any) -> dict[str, Any]:
    """Create response metadata from API response."""
    return {
        "model": getattr(response, "model", None),
        "finish_reason": getattr(choice, "finish_reason", None),
        "created": getattr(response, "created", None),
    }


def _create_ai_message(content: str, response_metadata: dict[str, Any]) -> AIMessage:
    """Create AIMessage with standard format."""
    return AIMessage(
        content=content,
        response_metadata=response_metadata,
        usage_metadata=None,  # Parallel doesn't return usage metadata
    )


def _create_stream_response_metadata(chunk: Any, choice: Any) -> dict[str, Any]:
    """Create response metadata for streaming chunks."""
    response_metadata = {}
    if hasattr(choice, "finish_reason") and choice.finish_reason is not None:
        response_metadata["finish_reason"] = str(choice.finish_reason)
    if hasattr(chunk, "model"):
        response_metadata["model"] = chunk.model
    return response_metadata


def _merge_consecutive_messages(messages: list[BaseMessage]) -> list[BaseMessage]:
    """Merge consecutive messages of the same type to satisfy API requirements.

    Parallel requires messages to alternate between user and assistant roles.
    This function merges consecutive messages of the same type.
    """
    if not messages:
        return messages

    merged: list[BaseMessage] = []
    current_content = []
    current_type = None

    for message in messages:
        message_type = type(message)

        if message_type == current_type:
            # Same type as previous, accumulate content
            current_content.append(str(message.content))
        else:
            # Different type, save previous and start new
            if current_type is not None and current_content:
                # Create merged message of the previous type
                merged_content = "\n\n".join(current_content)
                if current_type == SystemMessage:
                    merged.append(SystemMessage(content=merged_content))
                elif current_type == HumanMessage:
                    merged.append(HumanMessage(content=merged_content))
                elif current_type == AIMessage:
                    merged.append(AIMessage(content=merged_content))

            # Start new message
            current_type = message_type
            current_content = [str(message.content)]

    # Don't forget the last message
    if current_type is not None and current_content:
        merged_content = "\n\n".join(current_content)
        if current_type == SystemMessage:
            merged.append(SystemMessage(content=merged_content))
        elif current_type == HumanMessage:
            merged.append(HumanMessage(content=merged_content))
        elif current_type == AIMessage:
            merged.append(AIMessage(content=merged_content))

    return merged


class ChatParallelWeb(BaseChatModel):
    """Parallel Web chat model integration.

    This integration connects to Parallel's Chat API, which provides
    real-time web research capabilities through an OpenAI-compatible interface.

    Setup:
        Install ``langchain-parallel-web`` and set environment variable
        ``PARALLEL_AI_API_KEY``.

        .. code-block:: bash

            pip install -U langchain-parallel-web
            export PARALLEL_AI_API_KEY="your-api-key"

    Key init args — completion params:
        model: str
            Name of Parallel Web model to use. Defaults to "speed".
        temperature: Optional[float]
            Sampling temperature (ignored by Parallel).
        max_tokens: Optional[int]
            Max number of tokens to generate (ignored by Parallel).

    Key init args — client params:
        timeout: Optional[float]
            Timeout for requests.
        max_retries: int
            Max number of retries.
        api_key: Optional[str]
            Parallel API key. If not passed in will be read from env var
            PARALLEL_AI_API_KEY.
        base_url: str
            Base URL for Parallel API. Defaults to "https://api.parallel.ai".

    Instantiate:
        .. code-block:: python

            from langchain_parallel_web import ChatParallelWeb

            llm = ChatParallelWeb(
                model="speed",
                temperature=0.7,
                max_tokens=None,
                timeout=None,
                max_retries=2,
                # api_key="...",
                # other params...
            )

    Invoke:
        .. code-block:: python

            messages = [
                (
                    "system",
                    "You are a helpful assistant with access to real-time web "
                    "information."
                ),
                ("human", "What are the latest developments in AI?"),
            ]
            llm.invoke(messages)

    Stream:
        .. code-block:: python

            for chunk in llm.stream(messages):
                print(chunk.content, end="")

    Async:
        .. code-block:: python

            await llm.ainvoke(messages)

            # stream:
            async for chunk in llm.astream(messages):
                print(chunk.content, end="")

            # batch:
            await llm.abatch([messages])

    Token usage:
        .. code-block:: python

            ai_msg = llm.invoke(messages)
            ai_msg.usage_metadata

    Response metadata:
        .. code-block:: python

            ai_msg = llm.invoke(messages)
            ai_msg.response_metadata

    """

    model: str = Field(default="speed", alias="model_name")
    """The name of the model to use. Defaults to 'speed' for Parallel."""

    api_key: Optional[SecretStr] = Field(default=None)
    """Parallel API key. If not provided, will be read from
    PARALLEL_AI_API_KEY env var."""

    base_url: str = Field(default="https://api.parallel.ai")
    """Base URL for Parallel API."""

    temperature: Optional[float] = Field(default=None)
    """Sampling temperature (ignored by Parallel)."""

    max_tokens: Optional[int] = Field(default=None)
    """Max number of tokens to generate (ignored by Parallel)."""

    timeout: Optional[float] = Field(default=None)
    """Timeout for requests."""

    max_retries: int = Field(default=2)
    """Max number of retries."""

    # OpenAI-compatible parameters that are ignored by Parallel
    response_format: Optional[dict[str, Any]] = Field(default=None)
    """Response format (ignored by Parallel)."""

    tools: Optional[list[dict[str, Any]]] = Field(default=None)
    """Tools for function calling (ignored by Parallel)."""

    tool_choice: Optional[str] = Field(default=None)
    """Tool choice parameter (ignored by Parallel)."""

    stream_options: Optional[dict[str, Any]] = Field(default=None)
    """Stream options (ignored by Parallel)."""

    top_p: Optional[float] = Field(default=None)
    """Top-p sampling parameter (ignored by Parallel)."""

    frequency_penalty: Optional[float] = Field(default=None)
    """Frequency penalty (ignored by Parallel)."""

    presence_penalty: Optional[float] = Field(default=None)
    """Presence penalty (ignored by Parallel)."""

    logit_bias: Optional[dict[str, float]] = Field(default=None)
    """Logit bias (ignored by Parallel)."""

    seed: Optional[int] = Field(default=None)
    """Random seed (ignored by Parallel)."""

    user: Optional[str] = Field(default=None)
    """User identifier (ignored by Parallel)."""

    _client: Optional[openai.OpenAI] = None
    _async_client: Optional[openai.AsyncOpenAI] = None

    @model_validator(mode="after")
    def validate_environment(self) -> Self:
        """Validate that api key exists and initialize clients."""
        # Get API key from parameter or environment - this will raise if not found
        api_key_str = get_api_key(
            self.api_key.get_secret_value() if self.api_key else None
        )

        # Set the api_key field if it was loaded from environment
        if not self.api_key:
            self.api_key = SecretStr(api_key_str)

        # Initialize both sync and async OpenAI clients configured for Parallel
        self._client = get_openai_client(api_key_str, self.base_url)
        self._async_client = get_async_openai_client(api_key_str, self.base_url)
        return self

    @property
    def client(self) -> openai.OpenAI:
        """Get the sync OpenAI client, initializing if needed."""
        if self._client is None:
            msg = (
                "Client not initialized. Please ensure the model is properly validated."
            )
            raise ValueError(msg)
        return self._client

    @property
    def async_client(self) -> openai.AsyncOpenAI:
        """Get the async OpenAI client, initializing if needed."""
        if self._async_client is None:
            msg = (
                "Async client not initialized. "
                "Please ensure the model is properly validated."
            )
            raise ValueError(msg)
        return self._async_client

    @property
    def _llm_type(self) -> str:
        """Return type of chat model."""
        return "chat-parallel-web"

    @property
    def _identifying_params(self) -> dict[str, Any]:
        """Return a dictionary of identifying parameters."""
        return {
            "model": self.model,
            "base_url": self.base_url,
            "temperature": self.temperature,
            "max_tokens": self.max_tokens,
            "response_format": self.response_format,
            "tools": self.tools,
            "tool_choice": self.tool_choice,
        }

    @property
    def lc_secrets(self) -> dict[str, str]:
        """Return secrets for LangChain serialization."""
        return {"api_key": "PARALLEL_AI_API_KEY"}

    @property
    def lc_attributes(self) -> dict[str, Any]:
        """Return attributes for LangChain serialization."""
        attributes: dict[str, Any] = {}
        if self.base_url:
            attributes["base_url"] = self.base_url
        return attributes

    @classmethod
    def get_lc_namespace(cls) -> list[str]:
        """Get the namespace of the LangChain object."""
        return ["langchain_parallel_web", "chat_models"]

    @classmethod
    def is_lc_serializable(cls) -> bool:
        """Return whether this model can be serialized by LangChain."""
        return True

    @contextlib.contextmanager
    def _handle_errors(self) -> Iterator[None]:
        """Handle errors from Parallel API."""
        try:
            yield
        except AuthenticationError as e:
            msg = (
                f"Authentication failed with Parallel API. "
                f"Please check your API key: {e!s}"
            )
            raise ValueError(msg)
        except RateLimitError as e:
            msg = f"Rate limit exceeded for Parallel API. Please try again later: {e!s}"
            raise ValueError(msg)
        except Exception as e:
            msg = f"Error calling Parallel API: {e!s}"
            raise ValueError(msg)

    def _process_non_stream_response(self, response: Any) -> ChatResult:
        """Process a non-streaming response into a ChatResult."""
        choice = response.choices[0]
        content = choice.message.content or ""
        response_metadata = _create_response_metadata(response, choice)
        response_metadata["model"] = response_metadata["model"] or self.model

        message = _create_ai_message(content, response_metadata)
        generation = ChatGeneration(message=message)
        return ChatResult(generations=[generation])

    def _process_stream_chunk(
        self, chunk: Any, run_manager: Optional[CallbackManagerForLLMRun] = None
    ) -> Optional[ChatGenerationChunk]:
        """Process a streaming chunk into a ChatGenerationChunk."""
        if not (hasattr(chunk, "choices") and chunk.choices and len(chunk.choices) > 0):
            return None

        choice = chunk.choices[0]
        delta = choice.delta

        content = ""
        if hasattr(delta, "content") and delta.content is not None:
            content = delta.content

        response_metadata = _create_stream_response_metadata(chunk, choice)

        chunk_message = AIMessageChunk(
            content=content,
            response_metadata=response_metadata,
            usage_metadata=None,  # Parallel doesn't return usage metadata
        )

        if run_manager and content:
            run_manager.on_llm_new_token(content)

        return ChatGenerationChunk(message=chunk_message)

    async def _process_async_stream_chunk(
        self, chunk: Any, run_manager: Optional[AsyncCallbackManagerForLLMRun] = None
    ) -> Optional[ChatGenerationChunk]:
        """Process an async streaming chunk into a ChatGenerationChunk."""
        if not (chunk.choices and len(chunk.choices) > 0):
            return None

        choice = chunk.choices[0]
        delta = choice.delta

        content = ""
        if hasattr(delta, "content") and delta.content is not None:
            content = delta.content

        response_metadata = _create_stream_response_metadata(chunk, choice)

        chunk_message = AIMessageChunk(
            content=content,
            response_metadata=response_metadata,
            usage_metadata=None,  # Parallel doesn't return usage metadata
        )

        if run_manager and content:
            await run_manager.on_llm_new_token(content)

        return ChatGenerationChunk(message=chunk_message)

    def _generate(
        self,
        messages: list[BaseMessage],
        stop: Optional[list[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> ChatResult:
        """Generate a response using Parallel's chat API."""
        openai_messages = _prepare_messages(messages)

        with self._handle_errors():
            response = self.client.chat.completions.create(
                model=self.model,
                messages=cast(Any, openai_messages),
                stream=False,
                temperature=self.temperature,
                max_tokens=self.max_tokens,
                stop=stop,
            )

            return self._process_non_stream_response(response)

    def _stream(
        self,
        messages: list[BaseMessage],
        stop: Optional[list[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> Iterator[ChatGenerationChunk]:
        """Stream responses from Parallel's chat API."""
        openai_messages = _prepare_messages(messages)

        with self._handle_errors():
            stream = self.client.chat.completions.create(
                model=self.model,
                messages=cast(Any, openai_messages),
                stream=True,
                temperature=self.temperature,
                max_tokens=self.max_tokens,
                stop=stop,
            )

            for chunk in stream:
                chunk_result = self._process_stream_chunk(chunk, run_manager)
                if chunk_result is not None:
                    yield chunk_result

    async def _agenerate(
        self,
        messages: list[BaseMessage],
        stop: Optional[list[str]] = None,
        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> ChatResult:
        """Async generate a response using Parallel's chat API."""
        openai_messages = _prepare_messages(messages)

        with self._handle_errors():
            response = await self.async_client.chat.completions.create(
                model=self.model,
                messages=cast(Any, openai_messages),
                stream=False,
                temperature=self.temperature,
                max_tokens=self.max_tokens,
                stop=stop,
            )

            return self._process_non_stream_response(response)

    async def _astream(
        self,
        messages: list[BaseMessage],
        stop: Optional[list[str]] = None,
        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> AsyncIterator[ChatGenerationChunk]:
        """Async stream responses from Parallel's chat API."""
        openai_messages = _prepare_messages(messages)

        with self._handle_errors():
            stream = await self.async_client.chat.completions.create(
                model=self.model,
                messages=cast(Any, openai_messages),
                stream=True,
                temperature=self.temperature,
                max_tokens=self.max_tokens,
                stop=stop,
            )

            async for chunk in stream:
                chunk_result = await self._process_async_stream_chunk(
                    chunk, run_manager
                )
                if chunk_result is not None:
                    yield chunk_result
