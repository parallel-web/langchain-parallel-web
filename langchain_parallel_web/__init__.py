from importlib import metadata

from langchain_parallel_web.chat_models import ChatParallelWeb
from langchain_parallel_web.extract_tool import ParallelExtractTool
from langchain_parallel_web.search_tool import ParallelWebSearchTool

try:
    __version__ = metadata.version(__package__ or __name__)
except metadata.PackageNotFoundError:
    # Case where package metadata is not available.
    __version__ = ""
del metadata  # optional, avoids polluting the results of dir(__package__)

__all__ = [
    "ChatParallelWeb",
    "ParallelExtractTool",
    "ParallelWebSearchTool",
    "__version__",
]
