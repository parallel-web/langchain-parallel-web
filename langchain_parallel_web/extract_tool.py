"""Parallel AI Extract Tool for LangChain."""

from __future__ import annotations

from typing import Any, Optional

from langchain_core.callbacks import CallbackManagerForToolRun
from langchain_core.tools import BaseTool
from pydantic import BaseModel, Field, SecretStr, model_validator

from ._client import get_api_key, get_extract_client


class ParallelExtractInput(BaseModel):
    """Input schema for Parallel AI Extract Tool."""

    urls: list[str] = Field(description="List of URLs to extract content from")


class ParallelExtractTool(BaseTool):
    """Parallel AI Extract Tool.

    This tool extracts clean, structured content from web pages using the
    Parallel AI Extract API.

    Setup:
        Install ``langchain-parallel-web`` and set environment variable
        ``PARALLEL_AI_API_KEY``.

        .. code-block:: bash

            pip install -U langchain-parallel-web
            export PARALLEL_AI_API_KEY="your-api-key"

    Key init args:
        api_key: Optional[SecretStr]
            Parallel AI API key. If not provided, will be read from
            PARALLEL_AI_API_KEY env var.
        base_url: str
            Base URL for Parallel AI API. Defaults to "https://api.parallel.ai".
        max_chars_per_extract: Optional[int]
            Maximum characters per extracted result.

    Instantiation:
        .. code-block:: python

            from langchain_parallel_web import ParallelExtractTool

            # Basic instantiation
            tool = ParallelExtractTool()

            # With custom API key and parameters
            tool = ParallelExtractTool(
                api_key="your-api-key",
                max_chars_per_extract=5000
            )

    Invocation:
        .. code-block:: python

            # Extract content from URLs
            result = tool.invoke({
                "urls": [
                    "https://example.com/article1",
                    "https://example.com/article2"
                ]
            })

            # Result is a list of dicts with url, title, and content
            for item in result:
                print(f"Title: {item['title']}")
                print(f"URL: {item['url']}")
                print(f"Content: {item['content'][:200]}...")

    Response Format:
        Returns a list of dictionaries, each containing:
        - url: The URL that was extracted
        - title: Title of the webpage
        - content: Full extracted content as markdown
        - publish_date: Publish date if available (optional)
    """

    name: str = "parallel_extract"
    description: str = (
        "Extract clean, structured content from web pages. "
        "Input should be a list of URLs to extract content from. "
        "Returns extracted content formatted as markdown."
    )
    args_schema: type[BaseModel] = ParallelExtractInput

    api_key: Optional[SecretStr] = Field(default=None)
    """Parallel AI API key. If not provided, will be read from env var."""

    base_url: str = Field(default="https://api.parallel.ai")
    """Base URL for Parallel AI API."""

    max_chars_per_extract: Optional[int] = None
    """Maximum characters per extracted result."""

    @model_validator(mode="before")
    @classmethod
    def validate_environment(cls, values: dict) -> Any:
        """Validate the environment."""
        # Get API key from parameter or environment
        api_key = values.get("api_key")
        if isinstance(api_key, SecretStr):
            api_key_str: Optional[str] = api_key.get_secret_value()
        else:
            api_key_str = api_key

        # This will raise an error if API key is not found
        get_api_key(api_key_str)

        return values

    def _run(
        self,
        urls: list[str],
        run_manager: Optional[CallbackManagerForToolRun] = None,
    ) -> list[dict[str, Any]]:
        """Extract content from URLs.

        Args:
            urls: List of URLs to extract content from
            run_manager: Callback manager for the tool run

        Returns:
            List of dictionaries with extracted content
        """
        try:
            # Get API key
            api_key_str = get_api_key(
                self.api_key.get_secret_value() if self.api_key else None
            )

            # Initialize extract client
            client = get_extract_client(api_key_str, self.base_url)

            # Build full_content config if max_chars is specified
            full_content = None
            if self.max_chars_per_extract:
                full_content = {"max_characters": self.max_chars_per_extract}

            # Extract content from URLs
            extract_response = client.extract(
                urls=urls,
                full_content=full_content,
            )

            results = extract_response.get("results", [])
            errors = extract_response.get("errors", [])

            # Format results
            formatted_results = []
            for result in results:
                formatted_result = {
                    "url": result.get("url"),
                    "title": result.get("title"),
                    "content": result.get("full_content", ""),
                }

                # Add optional fields if present
                if "publish_date" in result:
                    formatted_result["publish_date"] = result["publish_date"]

                formatted_results.append(formatted_result)

            # If there were errors, add them to the results with error info
            formatted_results.extend(
                [
                    {
                        "url": error.get("url"),
                        "title": None,
                        "content": f"Error: {error.get('message', 'Unknown error')}",
                        "error_type": error.get("error_type"),
                    }
                    for error in errors
                ]
            )

            return formatted_results

        except Exception as e:
            msg = f"Error calling Parallel AI Extract API: {e!s}"
            raise ValueError(msg) from e

    async def _arun(
        self,
        urls: list[str],
        run_manager: Optional[CallbackManagerForToolRun] = None,
    ) -> list[dict[str, Any]]:
        """Extract content from URLs asynchronously.

        Args:
            urls: List of URLs to extract content from
            run_manager: Callback manager for the tool run

        Returns:
            List of dictionaries with extracted content
        """
        from ._client import get_async_extract_client

        try:
            # Get API key
            api_key_str = get_api_key(
                self.api_key.get_secret_value() if self.api_key else None
            )

            # Initialize async extract client
            client = get_async_extract_client(api_key_str, self.base_url)

            # Build full_content config if max_chars is specified
            full_content = None
            if self.max_chars_per_extract:
                full_content = {"max_characters": self.max_chars_per_extract}

            # Extract content from URLs
            extract_response = await client.extract(
                urls=urls,
                full_content=full_content,
            )

            results = extract_response.get("results", [])
            errors = extract_response.get("errors", [])

            # Format results
            formatted_results = []
            for result in results:
                formatted_result = {
                    "url": result.get("url"),
                    "title": result.get("title"),
                    "content": result.get("full_content", ""),
                }

                # Add optional fields if present
                if "publish_date" in result:
                    formatted_result["publish_date"] = result["publish_date"]

                formatted_results.append(formatted_result)

            # If there were errors, add them to the results with error info
            formatted_results.extend(
                [
                    {
                        "url": error.get("url"),
                        "title": None,
                        "content": f"Error: {error.get('message', 'Unknown error')}",
                        "error_type": error.get("error_type"),
                    }
                    for error in errors
                ]
            )

            return formatted_results

        except Exception as e:
            msg = f"Error calling Parallel AI Extract API: {e!s}"
            raise ValueError(msg) from e
