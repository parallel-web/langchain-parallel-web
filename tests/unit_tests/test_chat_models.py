"""Test chat model integration."""

from __future__ import annotations

from langchain_tests.unit_tests import ChatModelUnitTests

from langchain_parallel_web.chat_models import ChatParallelWeb


class TestChatParallelWebUnit(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> type[ChatParallelWeb]:
        return ChatParallelWeb

    @property
    def chat_model_params(self) -> dict:
        # These should be parameters used to initialize your integration for testing
        return {
            "model": "speed",
            "api_key": "test-api-key",
        }

    # Configure capabilities based on Parallel's Chat API features
    @property
    def has_tool_calling(self) -> bool:
        """Parallel Chat API tool calling support - currently not implemented."""
        return False

    @property
    def has_tool_choice(self) -> bool:
        """Parallel ignores tool choice parameter."""
        return False

    @property
    def has_structured_output(self) -> bool:
        """Parallel Chat API structured output support - currently not implemented.

        Currently not implemented in Parallel Chat API.
        """
        return False

    @property
    def supports_json_mode(self) -> bool:
        """Parallel ignores JSON mode parameter."""
        return False

    @property
    def returns_usage_metadata(self) -> bool:
        """Parallel Chat API does not currently return usage metadata."""
        return False

    @property
    def supports_anthropic_computer_use(self) -> bool:
        """Parallel Chat API does not support Anthropic computer use."""
        return False

    @property
    def supports_image_inputs(self) -> bool:
        """Parallel Chat API image input support - not confirmed."""
        return False

    @property
    def supports_image_urls(self) -> bool:
        """Parallel does not support image URLs."""
        return False

    @property
    def supports_pdf_inputs(self) -> bool:
        """Parallel does not support PDF inputs."""
        return False

    @property
    def supports_audio_inputs(self) -> bool:
        """Parallel Chat API does not support audio inputs."""
        return False

    @property
    def supports_video_inputs(self) -> bool:
        """Parallel Chat API does not support video inputs."""
        return False

    @property
    def supports_image_tool_message(self) -> bool:
        """Parallel does not support image tool messages."""
        return False

    @property
    def structured_output_kwargs(self) -> dict:
        """Additional kwargs for with_structured_output."""
        return {"method": "function_calling"}

    @property
    def supported_usage_metadata_details(self) -> dict:
        """Parallel supports basic usage metadata."""
        return {
            "invoke": [],
            "stream": [],
        }

    @property
    def enable_vcr_tests(self) -> bool:
        """Disable VCR tests for now."""
        return False

    @property
    def supports_system_messages(self) -> bool:
        """Parallel Chat API supports system messages via OpenAI interface.

        Supports system messages through OpenAI-compatible API.
        """
        return True

    @property
    def init_from_env_params(self) -> tuple[dict, dict, dict]:
        """Parameters for testing initialization from environment variables."""
        return (
            {
                "PARALLEL_AI_API_KEY": "test-env-api-key",
            },
            {
                "model": "speed",
            },
            {
                "api_key": "test-env-api-key",
            },
        )
