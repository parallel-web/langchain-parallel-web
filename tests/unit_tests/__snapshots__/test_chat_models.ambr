# serializer version: 1
# name: TestChatParallelWebUnit.test_serdes[serialized]
  dict({
    'id': list([
      'langchain_parallel_web',
      'chat_models',
      'ChatParallelWeb',
    ]),
    'kwargs': dict({
      'api_key': dict({
        'id': list([
          'PARALLEL_AI_API_KEY',
        ]),
        'lc': 1,
        'type': 'secret',
      }),
      'base_url': 'https://api.parallel.ai',
      'max_retries': 2,
      'max_tokens': 100,
      'model': 'speed',
      'temperature': 0.0,
      'timeout': 60.0,
    }),
    'lc': 1,
    'name': 'ChatParallelWeb',
    'type': 'constructor',
  })
# ---
